cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -695,6 +702,65 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
     }
   }
 
+  /**
+   * Delete the business metadata for the application and the programs.
+   */
+  private void deleteAppBusinessMetadata(Id.Application appId) {
+    // Remove metadata for the Application itself.
+    businessMds.removeProperties(appId);
+    businessMds.removeTags(appId);
+
+    // Remove business metadata for the programs of the Application
+    // TODO: Need to remove this we support prefix search of metadata type.
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    for (ProgramType programType : ProgramType.values()) {
+      deleteProgramBusinessMetadataForApp(programType, appId, appSpec);
+    }
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
+   */
+  private void deleteProgramBusinessMetadataForApp(ProgramType programType, Id.Application appId,
```

Comment:
This code is brittle. Everytime we add a new program type, we need to remember to add the program type here to delete. Why not delete the Business metadata based on a prefix key for app.